x-pack/plugins/enterprise_search/public/applications/app_search/components/settings/log_retention/utils/convert_log_retention.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  ELogRetentionOptions,
+  ILogRetention,
+  ILogRetentionPolicy,
+  ILogRetentionServer,
+  ILogRetentionServerPolicy,
+  ILogRetentionServerSettings,
+  ILogRetentionSettings,
+} from '../types';
+
+export const convertLogRetentionFromServerToClient = (
```

Comment:
I think we have a really unique opportunity in Kibana to clean up a lot of these `ServerToClient` transforms/conversions by handling them in our `server`/API code rather than our `public`/client-side code.

We do something very similar already in our [internal client API call](https://github.com/elastic/kibana/blob/5812217631d4a3dabbaddd687151b7b865e92c6f/x-pack/plugins/enterprise_search/server/lib/enterprise_search_config_api.ts#L64), for example.

I played around with this a bit and have a working diff for you:

https://github.com/JasonStoltz/kibana/compare/log-retention-logic...constancecchen:server-json-transform

Here's some screenshots of the final API output coming from Kibana:

<img width="495" alt="Screen Shot 2020-11-03 at 6 24 46 PM" src="https://user-images.githubusercontent.com/549407/98064549-7024db00-1e07-11eb-9f4e-b74262748bcf.png">
<img width="477" alt="Screen Shot 2020-11-03 at 6 53 04 PM" src="https://user-images.githubusercontent.com/549407/98064553-71560800-1e07-11eb-8d47-1d133add9381.png">

Would really love to hear what you think! Personally, I think it really helps separate concerns quite a bit.

Reply:
In any other system, I would be in favor of moving stuff like this to the server. 

In the case of Kibana, I am not in favor of it. I am strongly in favor of keeping our Kibana server API as thin as possible as primarily as pass through to Enterprise Search. It feels like we're monkey patching the Enterprise Search API rather than fixing it at its source.

Reply:
> It feels like we're monkey patching the Enterprise Search API rather than fixing it at its source.

Yes, but we're doing that anyway with these `ServerToClient` utilities. I'm simply proposing moving the currently existing to the `server/` folder.

> I am strongly in favor of keeping our Kibana server API as thin as possible as primarily as pass through to Enterprise Search.

I don't particularly see why our server API needs to be this way. This is also still a relatively thin action, IMO. I'm not sure what advantage we gain with that mindset and I think being overly prescriptive about what we do in our `server/` code is a wasted opportunity for cleanup / giving more freedom & control to our front-end devs.

Reply:
> I don't particularly see why our server API needs to be this way. This is also still a relatively thin action, IMO. I'm not sure what advantage we gain with that mindset and I think being overly prescriptive about what we do in our server/ code is a wasted opportunity for cleanup / giving more freedom & control to our front-end devs.

It's just one less place we have to look for logic. One less architectural piece to account for.

Also, if we just keep things as is, it makes it easier to just port code over from the old repository. This is one more step for porting views that we could easily cut out.

Reply:
> It's just one less place we have to look for logic. One less architectural piece to account for.

I'm not sold by that argument because `convertLogRetentionFromServerToClient` is also one more place we have to look for logic for. I had to open the util file and figure out what exactly the code was doing, etc.

If the concern is that it's a pretty disparate location, I also think there's a danger in that. I really don't think we should treat the `server/` folder as something to not look at or stay away from - it's integral for telemetry, for one - and I definitely don't want to be the only dev with familiarity with the code there. In general, I don't think there's any code or area where we should just say "Oh, we don't have to bother looking deeply here".

> Also, if we just keep things as is, it makes it easier to just port code over from the old repository. This is one more step for porting views that we could easily cut out.

Definitely a fair argument - for what it's worth I just did a quick grep in ent-search and it looks like these transforms are pretty recent and only occur here (LogRetention) and in the new Crawler folder (and that's something I'd be happy to chat w/ Byron about soon).

With that small scope in mind I think this is worth making this refactor now - it sets up a pattern and precedent that we _can_ use in the future but don't have to. I'm **definitely** not saying we should go into any of the existing API endpoints or logic files and try to fix all our various inconsistent casing during the migration, but that this allows us to do so after/as a piece of tech debt if we chose to.

I'm also primarily making this proposal now _because_ of the existing `convertServerToClient` that I think fits perfectly in `server/` - if this utility hadn't been here and the existing logic had just dealt with snake_cased json as-is, I wouldn't have proposed this / would have been fine to leave it as-is.

Reply:
Why not make this change at the source then, in ent-search?

Reply:
Nevermind, this is not worth a disagreement. Could you please make it a separate PR though, after this is merged?

Reply:
👍 I'm definitely good with this being separate, and after some thought I'd be fine leaving this off as post-migration tech debt if we'd rather stay focused for now.

> Why not make this change at the source then, in ent-search?

That's a good question - the only potential problem I can think of is one we might potentially see as more users move to Kibana: what happens when ent-search and kibana versions are out of sync (e.g. a user upgrades ent-search but not kibana)? Kibana will now start looking for JSON data/keys that don't exist or have changed, and we'll start seeing more bugs/errors as a result.

I'm slightly more inclined to prefer making small changes like casing transforms in the Kibana side as a result (less potential for version mismatch) and keep ent-search more stable.

Reply:
Great point, thanks 👍

Reply:
heavy 👍 for leaving this off as post-migration tech debt